File: android/mlkit-automl/app/src/main/java/com/google/firebase/codelab/mlkit/automl/ImageClassifier.kt
/**
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.firebase.codelab.mlkit.automl

import android.content.Context
import android.graphics.Bitmap
import android.os.SystemClock
import android.util.Log
import android.widget.Toast
import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.TaskCompletionSource
import com.google.firebase.ml.common.FirebaseMLException
import com.google.firebase.ml.common.modeldownload.FirebaseLocalModel
import com.google.firebase.ml.common.modeldownload.FirebaseModelManager
import com.google.firebase.ml.common.modeldownload.FirebaseRemoteModel
import com.google.firebase.ml.vision.FirebaseVision
import com.google.firebase.ml.vision.common.FirebaseVisionImage
import com.google.firebase.ml.vision.label.FirebaseVisionImageLabel
import com.google.firebase.ml.vision.label.FirebaseVisionImageLabeler
import com.google.firebase.ml.vision.label.FirebaseVisionOnDeviceAutoMLImageLabelerOptions
import java.io.IOException
import java.util.Locale

/** Classifies images with ML Kit AutoML.  */
class ImageClassifier
/** Initializes an `ImageClassifier`.  */
@Throws(FirebaseMLException::class)
internal constructor(context: Context) {

  /** MLKit AutoML Image Classifier  */
  private val labeler: FirebaseVisionImageLabeler?
  private var remoteModelDownloadSucceeded = false

  init {
    val remoteModel = FirebaseRemoteModel.Builder(REMOTE_MODEL_NAME).build()
    FirebaseModelManager.getInstance()
            .registerRemoteModel(remoteModel)

    FirebaseModelManager.getInstance()
            .registerLocalModel(
                    FirebaseLocalModel.Builder(LOCAL_MODEL_NAME)
                            .setAssetFilePath(LOCAL_MODEL_PATH)
                            .build()
            )

    val options = FirebaseVisionOnDeviceAutoMLImageLabelerOptions.Builder()
            .setConfidenceThreshold(CONFIDENCE_THRESHOLD)
            .setLocalModelName(LOCAL_MODEL_NAME)
            .setRemoteModelName(REMOTE_MODEL_NAME)
            .build()

    labeler = FirebaseVision.getInstance().getOnDeviceAutoMLImageLabeler(options)

    Toast.makeText(
            context,
            "Begin downloading the remote AutoML model.",
            Toast.LENGTH_SHORT
    ).show()

    // Track the remote model download progress.
    FirebaseModelManager.getInstance()
            .downloadRemoteModelIfNeeded(remoteModel)
            .addOnCompleteListener { task ->
              if (task.isSuccessful) {
                Toast.makeText(
                        context,
                        "Download remote AutoML model success.",
                        Toast.LENGTH_SHORT
                ).show()
                remoteModelDownloadSucceeded = true
              } else {
                val downloadingError = "Error downloading remote model."
                Log.e(TAG, downloadingError, task.exception)
                Toast.makeText(context, downloadingError, Toast.LENGTH_SHORT).show()
              }
            }

    Log.d(TAG, "Created a Firebase ML Kit AutoML Image Labeler.")
  }

  /** Classifies a frame from the preview stream.  */
  internal fun classifyFrame(bitmap: Bitmap): Task<String> {
    if (labeler == null) {
      Log.e(TAG, "Image classifier has not been initialized; Skipped.")
      val e = IllegalStateException("Uninitialized Classifier.")

      val completionSource = TaskCompletionSource<String>()
      completionSource.setException(e)
      return completionSource.task
    }

    val startTime = SystemClock.uptimeMillis()
    val image = FirebaseVisionImage.fromBitmap(bitmap)

    return labeler.processImage(image).continueWith { task ->
      val endTime = SystemClock.uptimeMillis()
      Log.d(TAG, "Time to run model inference: " + java.lang.Long.toString(endTime - startTime))

      val labelProbList = task.result

      // print the results
      var textToShow = "Source: " +
              (if (this.remoteModelDownloadSucceeded) "Remote" else "Local") +
              " model\n"
      textToShow += "Latency: " + java.lang.Long.toString(endTime - startTime) + "ms\n"
      textToShow += if (labelProbList.isNullOrEmpty())
        "No Result"
      else
        printTopKLabels(labelProbList)

      textToShow
    }
  }

  /** Closes labeler to release resources.  */
  internal fun close() {
    try {
      labeler?.close()
    } catch (e: IOException) {
      Log.e(TAG, "Unable to close the labeler instance", e)
    }

  }

  /** Prints top-K labels, to be shown in UI as the results.  */
  private val printTopKLabels: (List<FirebaseVisionImageLabel>) -> String = {
    it.joinToString(
            separator = "\n",
            limit = RESULTS_TO_SHOW
    ) { label ->
      String.format(Locale.getDefault(), "Label: %s, Confidence: %4.2f", label.text, label.confidence)
    }
  }

  companion object {

    /** Tag for the [Log].  */
    private const val TAG = "MLKitAutoMLCodelab"

    /** Name of the local model file stored in Assets.  */
    private const val LOCAL_MODEL_NAME = "automl_image_labeling_model"

    /** Path of local model file stored in Assets.  */
    private const val LOCAL_MODEL_PATH = "automl/manifest.json"

    /** Name of the remote model in Firebase ML Kit server.  */
    private const val REMOTE_MODEL_NAME = "mlkit_flowers"

    /** Number of results to show in the UI.  */
    private const val RESULTS_TO_SHOW = 3

    /** Min probability to classify the given image as belong to a category.  */
    private const val CONFIDENCE_THRESHOLD = 0.6f
  }
}

File: ios/mlkit-automl/MLVisionExample/ImageClassifier.swift
//
//  Copyright 2019 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//  https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import Foundation
import FirebaseMLCommon
import FirebaseMLVisionAutoML

class ImageClassifer {

  private let vision = Vision.vision()
  private let modelManager = ModelManager.modelManager()
  private let autoMLOnDeviceLabeler: VisionImageLabeler

  private let localModel: LocalModel? = {
    guard let localModelFilePath = Bundle.main.path(
      forResource: Constant.autoMLManifestFileName,
      ofType: Constant.autoMLManifestFileType,
      inDirectory: Constant.autoMLManifestFolder
      ) else {
        print("Failed to find AutoML local model manifest file.")
        return nil
    }

    return LocalModel(name: Constant.localAutoMLModelName, path: localModelFilePath)
  }()

  private let remoteModel: RemoteModel = {
    let initialConditions = ModelDownloadConditions()
    let updateConditions = ModelDownloadConditions(
      allowsCellularAccess: false,
      allowsBackgroundDownloading: true
    )
    return RemoteModel(
      name: Constant.remoteAutoMLModelName,
      allowsModelUpdates: true,
      initialConditions: initialConditions,
      updateConditions: updateConditions
    )
  }()

  init() {
    // Load the remote AutoML model.
    modelManager.register(remoteModel)
    modelManager.download(remoteModel)

    // Load the local AutoML model.
    var localModelName: String?
    if let model = localModel {
      modelManager.register(model)
      localModelName = Constant.localAutoMLModelName
    }

    // Create AutoML image labeler.
    let options = VisionOnDeviceAutoMLImageLabelerOptions(
      remoteModelName: Constant.remoteAutoMLModelName,
      localModelName: localModelName
    )
    options.confidenceThreshold = Constant.labelConfidenceThreshold
    autoMLOnDeviceLabeler = vision.onDeviceAutoMLImageLabeler(options: options)

    // Set up to get notified when remote model download succeeded.
    setupRemoteModelDownloadNotification()
  }
}

// MARK: - Private

extension ImageClassifer {

  /// Set up receiver to get notified about model download progress.
  private func setupRemoteModelDownloadNotification() {
    NotificationCenter.default.addObserver(
      forName: .firebaseMLModelDownloadDidSucceed,
      object: nil,
      queue: OperationQueue.main
    ) { _ in
      print("Sucessfully downloaded AutoML remote model")
    }

    NotificationCenter.default.addObserver(
      forName: .firebaseMLModelDownloadDidFail,
      object: nil,
      queue: OperationQueue.main
    ) { _ in
      print("Error: AutoML remote model download failed. Check if Constant.remoteAutoMLModelName",
            "matches with the model name you published in the Firebase Console.")
    }
  }

}

// MARK: - Classify image

extension ImageClassifer {

  /// Classify the given UIImage instance. This method is useful to classify still images.
  func classifyImage(_ image: UIImage, completionHandler: @escaping ImageClassificationCompletion) {
    // Rotate the image so that its imageOrientation is always "up"
    guard let rotatedImage = image.imageOrientedUp() else {
      completionHandler(nil, ClassificationError.invalidInput)
      return
    }

    // Initialize a VisionImage object with the rotated image.
    let visionImage = VisionImage(image: rotatedImage)

    // Feed the image to ML Kit AutoML SDK.
    classifyVisionImage(visionImage, completionHandler: completionHandler)
  }

  /// Classify the given CMSampleBuffer instance.
  /// This method is useful to classify frames of video streams.
  func classifySampleBuffer(
    _ sampleBuffer: CMSampleBuffer,
    isUsingFrontCamera: Bool,
    completionHandler: @escaping ImageClassificationCompletion
  ) {

    let visionImage = VisionImage(buffer: sampleBuffer)
    let metadata = VisionImageMetadata()
    let orientation = UIUtilities.imageOrientation(
      fromDevicePosition: isUsingFrontCamera ? .front : .back
    )
    metadata.orientation = UIUtilities.visionImageOrientation(from: orientation)
    visionImage.metadata = metadata

    // Feed the image to ML Kit AutoML SDK.
    classifyVisionImage(visionImage, completionHandler: completionHandler)
  }

  /// Classify a VisionImage instance. This private method provides actual implementation of
  /// the image classification logic and is called by the public "classify" methods.
  private func classifyVisionImage(
    _ visionImage: VisionImage,
    completionHandler: @escaping ImageClassificationCompletion
  ) {

    // Return error if AutoML local model is not available.
    guard localModel != nil else {
      completionHandler(nil, ClassificationError.localModelNotAvailable)
      return
    }

    // Indicate whether the remote or local model is used.
    // Note: in most common cases, once a remote model is downloaded it will be used. However, in
    // very rare cases, the model itself might not be valid, and thus the local model is used. In
    // addition, since model download failures can be transient, and model download can also be
    // triggered in the background during inference, it is possible that a remote model is used
    // even if the first download fails.
    let isRemoteModelDownloaded = modelManager.isRemoteModelDownloaded(remoteModel)
    var result = "Source: " + (isRemoteModelDownloaded ? "Remote" : "Local") + " model\n"

    let startTime = DispatchTime.now()

    autoMLOnDeviceLabeler.process(visionImage) { detectedLabels, error in
      guard error == nil else {
        completionHandler(nil, error)
        return
      }

      // Measure inference latency and format it to show to user.
      let endTime = DispatchTime.now()
      let nanoTime = endTime.uptimeNanoseconds - startTime.uptimeNanoseconds
      let latencyInMs = round(Double(nanoTime) / 1_000_000)
      result += "Latency: \(latencyInMs)ms\n"

      // Format detection result to show to user.
      result += detectedLabels?.map { label -> String in
            return "Label: \(label.text), Confidence: \(label.confidence ?? 0)"
      }.joined(separator: "\n") ?? "No Result"

      completionHandler(result, nil)
    }
  }
}

// MARK: - Constants and types

private enum Constant {
  /// Definition of AutoML local model
  static let localAutoMLModelName = "automl_image_labeling_model"
  static let autoMLManifestFileName = "manifest"
  static let autoMLManifestFileType = "json"
  static let autoMLManifestFolder = "automl"

  /// Definition of AutoML remote model.
  static let remoteAutoMLModelName = "mlkit_flowers"

  /// Config for AutoML Image Labeler classification task.
  static let labelConfidenceThreshold: Float = 0.6
}

private enum ClassificationError: Error {
  case invalidInput
  case localModelNotAvailable
}

extension ClassificationError: LocalizedError {
  public var errorDescription: String? {
    switch self {
      case .invalidInput:
        return "Invalid input image."
      case .localModelNotAvailable:
        return "AutoML local model is not available. Please check if you have " +
          "downloaded and added the model files to this project."
    }
  }
}

typealias ImageClassificationCompletion = (String?, Error?) -> Void
